### ISSUE TITLE
`createModel`+`model.assign` with `strictFunctionTypes` enabled causes type errors

### ISSUE THREAD

**Issue by with-heart** [statelyai/xstate#2122]
**Description**
With `strictFunctionTypes` (or `strict`) compiler option enabled in `tsconfig.json`, when I use `model.assign` as an action for the machine, I receive type errors.

The errors seem to be around incompatibilities with the `exec` key, though as is usual with ts, the errors are quite dense and it's hard to follow what it's complaining about exactly.

**Expected Result**
No errors. The machines I've built with this error work fine, the types just aren't lining up for some reason. They also work fine with regular `assign`, it only becomes an issue with `model.assign`.

**Actual Result**
Different situations depending on how I call `createMachine`:

- `createMachine<typeof model>`, error on the `context` key: https://codesandbox.io/s/amazing-nash-h7g0n?file=/src/machine.ts

For some reason in this example, ts fails to understand what the type of `context` is in the definition. Hover over the `context` key, it shows `unknown`.

This setup works fine if I use regular `assign`: https://codesandbox.io/s/hungry-dubinsky-q5gli?file=/src/machine.ts:544-545

- `createMachine<ModelContextFrom<typeof model>, ModelEventsFrom<typeof model>>`, error on the individual actions: https://codesandbox.io/s/eager-borg-edlnb?file=/src/machine.ts

This setup also works fine if I use regular `assign`: https://codesandbox.io/s/condescending-minsky-djs7j?file=/src/machine.ts

**Comment by Andarist** [comment]
> The errors seem to be around incompatibilities with the exec key, though as is usual with ts, the errors are quite dense and it's hard to follow what it's complaining about exactly.

`exec` is a function and as such, this is a contravariant position for TS. This has an impact on assignability rules in TS (for good reasons) but it also impacts the TS inference algorithm. I'm not quite sure *why* this happens here - but the `event` parameter is somewhat computed to `never` here and, well, that ain't assignable to the specified `TEvent`. This is problematic because here we really expect the full `TEvent` to be inferred. 

Fun fact - if you declare your `increment`/`decrement` outside of the `createMachine` call (and let the declaration inference kick in) then you won't experience this problem.

And also note - we plan to remove `exec` property from the types, since in our situation we want to treat actions as bivariant. I've confirmed that this would solve your issue.

> For some reason in this example, ts fails to understand what the type of context is in the definition. Hover over the context key, it shows unknown.

This is just because the first overload fails to match, so it tries the second one which is not intended to be used with the model. This results in very cryptic errors like this 😬

**Comment by Andarist** [comment]
This is related to the model API that is already deprecated for some time. Therefore I think it's OK to close this issue. If you have any typing issues with the main `createMachine` API - let me know and I will try to assist you as much as I can.
